internal/middlewares/rateLimiter.go

```diff
@@ -0,0 +1,54 @@
+package middleware
+
+import (
+	"first-api/internal/ratelimiter"
+	"first-api/pkg/cache"
+	"fmt"
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+func RateLimitMiddleware(c *gin.Context, redisClient cache.UserCache, tb *ratelimiter.TokenBucket) {
+	// Check if the Token Bucket has enough tokens
+	if tb.IsTotalRequestAllowed(redisClient) {
+		// Perform the request
+		// check for per client basis request
+		clientID := GetClientIdentity(c)
+		clientTB := ratelimiter.GetClientBucket(clientID, "user", redisClient)
+		if clientTB == nil {
+			fmt.Println("failed to retreive client bucket from redis server")
+			c.AbortWithStatusJSON(http.StatusBadGateway, gin.H{
+				"message": "Something went wrong try again later ",
+			})
+			return
+		}
+		fmt.Println("clientID:", clientID)
+		if !clientTB.IsClientRequestAllowed(1) {
+			fmt.Println("too many request by ", clientID)
+			c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
+				"message": "Too many requests , try again after sometimes!",
+			})
+			return
+		}
+		err := ratelimiter.UpdateClientBucketInRedis(redisClient)
+		if err != nil {
+			fmt.Println("Error:", err)
+			c.AbortWithStatusJSON(http.StatusBadGateway, gin.H{
+				"message": err,
+			})
+		}
+		c.Next()
+	} else {
+		// Rate limit reached, reject the request
+		c.AbortWithStatus(http.StatusTooManyRequests)
+	}
+}
+
+func GetClientIdentity(c *gin.Context) string {
+	ip := c.ClientIP()
+	url := c.Request.URL.Path
```

Comment:
Endpoints of type /v1/user/:id can have different url paths (/v1/user/1; /v1/user/2). Due to this the GetClientIdentity will return different result for each variant of path which shouldn't be the case. So it's better to get the pattern from the path and map each endpoint with its rate limit rules.

Reply:
@anartify  Thanks for pointing out but this won't be the case in the actual production environment and the client identity will be same for all these calls , it's just in the test environment or postman that when we make a call the url is set as the clientip and we get different identity for each call